### PR TITLE
Qt: Correct 'Sort by Progress' for seeding torrents

### PR DESCRIPTION
**Pull request by OscarCunningham**
This is for issue #3796.

I don't like how complicated this turned out, but there are a lot of cases since each torrent can have either a normal limit, a limit of 0, or no limit at all.

The similar function `compareRatio` also accounts for the possibility that `ratio()` can be infinite, but I don't think this is actually possible because it could only result from a `downloadedBytes` of 0 which can't happen if you're seeding. Even seeding your own torrent the `downloadedBytes` is still positive. (If infinty is possible then we should probably also take care of `TR_RATIO_NA` in both functions.)

This commit also introduces some `-Wfloat-equal` warnings, but actually I think I'm right and the warning is stupid. If you really want to avoid the warning then I'll rearrange the code a bit to deal with it.

### CODE REVIEW

**Review by ckerr — COMMENTED** [review]
Thanks for this PR! You're right about all the different tests that need to happen here :dizzy:

File: qt/Torrent.cc [hunk 1]
```diff
@@ -54,27 +54,48 @@ bool Torrent::includesTracker(QString const& sitename) const
     return std::binary_search(std::begin(sitenames_), std::end(sitenames_), sitename);
 }
 
-int Torrent::compareSeedRatio(Torrent const& that) const
+int Torrent::compareSeedProgress(Torrent const& that) const
 {
-    auto const a = getSeedRatioLimit();
-    auto const b = that.getSeedRatioLimit();
+    auto const a_ratio = ratio();
+    auto const b_ratio = that.ratio();
```
[the remark this answers is not in the document]
Comment (OscarCunningham) [reply qt/Torrent.cc]: Done

File: qt/Torrent.cc [hunk 2]
```diff
@@ -54,27 +54,48 @@ bool Torrent::includesTracker(QString const& sitename) const
...
+    {
+        return b_ratio < *b_ratio_limit ? 1 : -1;
+    }
+
+    if (!b_ratio_limit)
+    {
+        return a_ratio < *a_ratio_limit ? -1 : 1;
+    }
+
+    if (*a_ratio_limit == 0 && *b_ratio_limit == 0)
```
[the remark this answers is not in the document]
Comment (OscarCunningham) [reply qt/Torrent.cc]: Done

**Review by ckerr — APPROVED** [review]

**Comment by Coeur** [comment]
>This commit also introduces some `-Wfloat-equal` warnings, but actually I think I'm right and the warning is stupid.

I agree on this, and I was fighting against that warning in the past in #4400 which got 46 comments for a 4-lines fix and was rejected in the end.

**Pull request merged**